contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol

```diff
@@ -0,0 +1,159 @@
+     * @param _decimals L2 ERC20 decimals
+     * @param _name L2 ERC20 name
+     * @param _symbol L2 ERC20 symbol
+     */
+    constructor(
+        address _l2CrossDomainMessenger,
+        uint8 _decimals,
+        string memory _name,
+        string memory _symbol
+    )
+        public
+        OVM_CrossDomainEnabled(_l2CrossDomainMessenger)
+        UniswapV2ERC20(_decimals, _name, _symbol)
+    {}
+
+    // Should we be inheriting this initialization logic?
+    /**
+     * @dev Initialize this gateway with the L1 gateway address
+     * The assumed flow is that this contract is deployed, then the L1 gateway,
+     * then init is called on L2.
+     * @param _l1ERC20Gateway Address of the corresponding L1 gateway deployed to the main chain
+     */
+    function init(
+        iOVM_L1ERC20Gateway _l1ERC20Gateway
+    )
+        public
+    {
+        require(address(l1ERC20Gateway) == address(0), "Contract has already been initialized");
+
+        l1ERC20Gateway = _l1ERC20Gateway;
+        
+        emit Initialized(l1ERC20Gateway);
+    }
+
+    /**********************
+     * Function Modifiers *
+     **********************/
+
+    modifier onlyInitialized() {
+        require(address(l1ERC20Gateway) != address(0), "Contract has not yet been initialized");
+        _;
+    }
+
+    /***************
+     * Withdrawing *
+     ***************/
+
+    /**
+     * @dev initiate a withdraw of some ERC20 to the caller's account on L1
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function withdraw(
+        uint _amount
+    )
+        external
+        override
+        onlyInitialized()
+    {
+        _initiateWithdrawal(msg.sender, _amount);
+    }
+
+    /**
+     * @dev initiate a withdraw of some ERC20 to a recipient's account on L1
+     * @param _to L1 adress to credit the withdrawal to
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function withdrawTo(address _to, uint _amount) external override onlyInitialized() {
+        _initiateWithdrawal(_to, _amount);
+    }
+
+    /**
+     * @dev Performs the logic for deposits by storing the ERC20 and informing the L2 ERC20 Gateway of the deposit.
+     *
+     * @param _to Account to give the withdrawal to on L1
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function _initiateWithdrawal(address _to, uint _amount) internal {
+        // burn L2 funds so they can't be used more on L2
+        _burn(msg.sender, _amount);
+
+        // Construct calldata for l1ERC20Gateway.finalizeWithdrawal(_to, _amount)
+        bytes memory data = abi.encodeWithSelector(
+            iOVM_L1ERC20Gateway.finalizeWithdrawal.selector,
+            _to,
+            _amount
+        );
+
+        // Send message up to L1 gateway
+        sendCrossDomainMessage(
+            address(l1ERC20Gateway),
+            data,
+            8999999 // TODO: meter and set with some buffer (actually I think its currently unused in this direction)
```

Comment:
Would be good to inherit this value from the EM. And maybe do something like `ExecutionManager.GAS_LIMIT - 1` or something